tests/api/runtimes/base.py

```diff
@@ -0,0 +1,207 @@
+            f"Setting up test {self.__class__.__name__}::{method.__name__}"
+        )
+
+        self.custom_setup()
+
+        self._logger.info(
+            f"Finished setting up test {self.__class__.__name__}::{method.__name__}"
+        )
+
+    def custom_setup(self):
+        pass
+
+    def _generate_task(self):
+        return new_task(
+            name=self.name, project=self.project, artifact_path=self.artifact_path
+        )
+
+    def _mock_create_namespaced_pod(self):
+        def _generate_pod(namespace, pod):
+            terminated_container_state = client.V1ContainerStateTerminated(
+                finished_at=datetime.now(timezone.utc), exit_code=0
+            )
+            container_state = client.V1ContainerState(
+                terminated=terminated_container_state
+            )
+            container_status = client.V1ContainerStatus(
+                state=container_state,
+                image=self.image_name,
+                image_id="must-provide-image-id",
+                name="must-provide-name",
+                ready=True,
+                restart_count=0,
+            )
+            status = client.V1PodStatus(
+                phase=PodPhases.succeeded, container_statuses=[container_status]
+            )
+            response_pod = deepcopy(pod)
+            response_pod.status = status
+            response_pod.metadata.name = "test-pod"
+            response_pod.metadata.namespace = namespace
+            return response_pod
+
+        get_k8s().v1api.create_namespaced_pod = unittest.mock.Mock(
+            side_effect=_generate_pod
+        )
+
+    def _execute_run(self, runtime, **kwargs):
+        # Reset the mock, so that when checking is create_pod was called, no leftovers are there (in case running
+        # multiple runs in the same test)
+        get_k8s().v1api.create_namespaced_pod.reset_mock()
+
+        runtime.run(
+            name=self.name,
+            project=self.project,
+            artifact_path=self.artifact_path,
+            **kwargs,
+        )
+
+    def _assert_labels_exist_in_pod_creation(self, labels: dict):
+        expected_labels = {
+            "mlrun/class": "job",
+            "mlrun/name": self.name,
+            "mlrun/project": self.project,
+            "mlrun/tag": "latest",
+        }
+
+        for key in expected_labels:
+            assert labels[key] == expected_labels[key]
+
+    def _assert_function_config_as_expected(
+        self,
+        config,
+        expected_params,
+        expected_inputs,
+        expected_hyper_params,
+        expected_secrets,
+    ):
+        function_metadata = config["metadata"]
+        assert function_metadata["name"] == self.name
+        assert function_metadata["project"] == self.project
+
+        function_spec = config["spec"]
+        assert function_spec["output_path"] == self.artifact_path
+        if expected_params:
+            assert function_spec["parameters"] == expected_params
```

Comment:
IIRC this is a dict, use deepdiff, same for other assertions when it's not a simple type assertion